src/core/spectrum_image.cpp

```diff
@@ -0,0 +1,671 @@
+                        {
+                            this[0].GetRealValueByLinearInterpolationNoBoundsCheckImage(ii_phys, jj_phys, current_interpolated_value);
+                            //MyDebugPrint("%g %g\n",current_interpolated_value,other_image[0].real_values[address_in_other_image]);
+                            cc_numerator_dist.AddSampleValue(current_interpolated_value * other_image[current_image].real_values[address_in_other_image]);
+                            cc_denom_other_dist.AddSampleValue(pow(other_image[0].real_values[address_in_other_image], 2)); // potential optimization: since other_image is not being rotated, we should only need to compute this quantity once, not for every potential rotation
+                            cc_denom_self_dist.AddSampleValue(pow(current_interpolated_value, 2));
+                        }
+                    }
+                    address_in_other_image++;
+                } // i
+                address_in_other_image += other_image[0].padding_jump_value;
+            } // end of loop over other (reference) image
+        } // end of loop over array of images
+
+        current_cc = cc_numerator_dist.GetSampleSum( ) / sqrt(cc_denom_other_dist.GetSampleSum( ) * cc_denom_self_dist.GetSampleSum( ));
+
+        if ( current_cc > best_cc ) {
+            best_cc       = current_cc;
+            best_rotation = current_rotation;
+        }
+
+        // Increment the rotation
+        current_rotation += search_step_size;
+
+    } // end of loop over rotations
+
+    return best_rotation;
+}
+
+// This function assumes the image is in real space and generates the power spectrum
+void SpectrumImage::GeneratePowerspectrum(CTF ctf_to_apply) {
+    MyDebugAssertTrue(is_in_memory, "Memory not allocated");
+    MyDebugAssertTrue(is_in_real_space == true, "image not in real space");
+    MyDebugAssertTrue(logical_z_dimension == 1, "Volumes not supported");
+
+    int         i, j;
+    float       j_logi, j_logi_sq, i_logi, i_logi_sq;
+    float       current_spatial_frequency_squared;
+    long        address                     = 0;
+    const float inverse_logical_x_dimension = 1.0 / float(logical_x_dimension);
+    const float inverse_logical_y_dimension = 1.0 / float(logical_y_dimension);
+    float       current_azimuth;
+    float       current_ctf_value;
+
+    for ( j = 0; j < logical_y_dimension; j++ ) {
+        address   = j * (padding_jump_value + 2 * physical_address_of_box_center_x);
+        j_logi    = float(j - physical_address_of_box_center_y) * inverse_logical_y_dimension;
+        j_logi_sq = powf(j_logi, 2);
+        for ( i = 0; i < logical_x_dimension; i++ ) {
+            i_logi    = float(i - physical_address_of_box_center_x) * inverse_logical_x_dimension;
+            i_logi_sq = powf(i_logi, 2);
+
+            // Where are we?
+            current_spatial_frequency_squared = j_logi_sq + i_logi_sq;
+
+            current_azimuth   = atan2f(j_logi, i_logi);
```

Comment:
Many other places in the code have a logical to check for zero frequency and return zero which I assume is to avoid undefined behavior for atan2f under some implementations?